tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,942 @@
+			return srv_conn;
+	}
+
+	return NULL;
+}
+
+/**
+ * Same as @tfw_sched_ratio_sched_sg_conn(), but schedule a connection
+ * to a specific server in a group.
+ */
+static TfwSrvConn *
+tfw_sched_ratio_sched_srv_conn(TfwMsg *msg, TfwServer *srv)
+{
+        int skipnip = 1, nipconn = 0;
+        TfwRatioSrvDesc *srvdesc = srv->sched_data;
+        TfwSrvConn *srv_conn;
+
+        /*
+	 * For @srv without connections @srvdesc will be NULL. Normally,
+	 * it doesn't happen in real life, but unit tests check this case.
+	 */
+        if (unlikely(!srvdesc))
+                return NULL;
+rerun:
+        if ((srv_conn = __sched_srv(srvdesc, skipnip, &nipconn)))
+                return srv_conn;
+
+        if (skipnip && nipconn) {
+                skipnip = 0;
+                goto rerun;
+        }
+
+        return NULL;
+}
+
+/**
+ * On each subsequent call the function returns the next available
+ * connection to one of the servers in the group. Connections to a
+ * server are rotated in pure round-robin fashion.
+ *
+ * A server is chosen according to its current weight that can be
+ * either static or dynamic. Servers with greater weight are chosen
+ * more often than servers with lesser weight.
+ *
+ * Dead connections and servers w/o live connections are skipped.
+ * Initially, connections with non-idempotent requests are also skipped
+ * in attempt to increase throughput. However, if all live connections
+ * contain a non-idempotent request, then re-run the algorithm and get
+ * the first live connection they way it is usually done.
+ *
+ * Ratio scheduler must be the fastest scheduler. Also, it's essential
+ * to maintain a completely fair distribution of requests to servers
+ * according to servers weights.
+ */
+static TfwSrvConn *
+tfw_sched_ratio_sched_sg_conn(TfwMsg *msg, TfwSrvGroup *sg)
+{
+	size_t srv_tried_n = 0;
+        int skipnip = 1, nipconn = 0;
+	TfwRatioPool *rpool = sg->sched_data;
+	TfwRatioSrvDesc *srvdesc, *srvdesc_last = NULL;
+	TfwSrvConn *srv_conn;
+	TfwRatio *ratio;
+
+	BUG_ON(!rpool);
+
+	rcu_read_lock();
+	ratio = rcu_dereference(rpool->ratio);
+	BUG_ON(!ratio);
+rerun:
+	/*
+	 * Try each server in a group. Attempt to schedule a connection
+	 * to a server that doesn't fall under a set of restrictions.
+	 *
+	 * FIXME: The way the algorithm works, same server may be chosen
+	 * multiple times in a row, even if that's the server where all
+	 * connections were under restrictions for one reason or another.
+	 * The idea is that the conditions for server's connections may
+	 * change any time, and so the next time one or more connections
+	 * to the same server will not be restricted.
+	 * Perhaps, though, it makes sense to skip these servers that
+	 * were restricted, and go directly to the next server. Getting
+	 * the next server reqires a lock, so perhaps it makes sense to
+	 * to skip these repetitive servers while under the lock.
+	 */
+	while (srv_tried_n < ratio->srv_n) {
```

Comment:
Actually, condition `while (srv_tried_n < ratio->srv_n)` will never mean, that we try all possible servers. And you described that in a comment above. Better to rename `srv_tried_n` to `tries`:
```
tries = ratio->srv_n + 1;
while(--tries) {
    ...
```
At least, this code doesn't lie that it tries all the server. Even more: nature of this counter is almost the same as for `tries` counter in hash scheduler.